flexmeasures/data/queries/utils.py

```diff
@@ -36,54 +36,68 @@ def create_beliefs_query(
     return query
 
 
-def add_user_source_filter(
-    cls: "ts.TimedValue", query: Query, user_source_ids: Union[int, List[int]]
-) -> Query:
-    """Add filter to the query to search only through user data from the specified user sources.
+def get_source_criteria(
+    cls: "Type[ts.TimedValue]",
+    user_source_ids: Union[int, List[int]],
+    source_types: List[str],
+    exclude_source_types: List[str],
+) -> List[bool]:
+    source_criteria: List[bool] = []
+    if user_source_ids is not None:
+        source_criteria.append(user_source_criterion(cls, user_source_ids))
+    if source_types is not None:
+        if user_source_ids and "user" not in source_types:
+            source_types.append("user")
+        source_criteria.append(source_type_criterion(source_types))
+    if exclude_source_types is not None:
+        if user_source_ids and "user" in exclude_source_types:
+            exclude_source_types.remove("user")
+        source_criteria.append(source_type_exclusion_criterion(exclude_source_types))
+    return source_criteria
+
+
+def user_source_criterion(
+    cls: "Type[ts.TimedValue]",
+    user_source_ids: Union[int, List[int]],
+) -> bool:
```

Comment:
Weird that we have to write `bool` here, because if I understand this correctly, we pass a WHERE part of an sqlalchemy query. Wouldn't it be better to not use a type hint here if there is no better way?

Reply:
I found out that the actual type of these parts of an SQLAlchemy query is `sqlalchemy.sql.elements.BinaryExpression`.